Add intrinsic support for the SVE multi-vector Advanced Encryption Standard (AES) instructions

amilendra opened the pull request:
---
name: Pull request
about: Technical issues, document format problems, bugs in scripts or feature proposal.

---

<!-- SPDX-FileCopyrightText: Copyright 2021-2022 Arm Limited and/or its affiliates <open-source-office@arm.com> -->
<!-- SPDX-License-Identifier: Apache-2.0 -->

**Thank you for submitting a pull request!**

If this PR is about a bugfix:

Please use the bugfix label and make sure to go through the checklist below.

If this PR is about a proposal:

We are looking forward to evaluate your proposal, and if possible to
make it part of the Arm C Language Extension (ACLE) specifications.

We would like to encourage you reading through the [contribution
guidelines](https://github.com/ARM-software/acle/blob/main/CONTRIBUTING.md), in particular the section on [submitting
a proposal](https://github.com/ARM-software/acle/blob/main/CONTRIBUTING.md#proposals-for-new-content).

Please use the proposal label.

As for any pull request, please make sure to go through the below
checklist.

Checklist: (mark with ``X`` those which apply)

* [X] If an issue reporting the bug exists, I have mentioned it in the
      PR (do not bother creating the issue if all you want to do is
      fixing the bug yourself).
* [X] I have added/updated the `SPDX-FileCopyrightText` lines on top
      of any file I have edited. Format is `SPDX-FileCopyrightText:
      Copyright {year} {entity or name} <{contact informations}>`
      (Please update existing copyright lines if applicable. You can
      specify year ranges with hyphen , as in `2017-2019`, and use
      commas to separate gaps, as in `2018-2020, 2022`).
* [X] I have updated the `Copyright` section of the sources of the
      specification I have edited (this will show up in the text
      rendered in the PDF and other output format supported). The
      format is the same described in the previous item.
* [X] I have run the CI scripts (if applicable, as they might be
      tricky to set up on non-*nix machines). The sequence can be
      found in the [contribution
      guidelines](https://github.com/ARM-software/acle/blob/main/CONTRIBUTING.md#continuous-integration). Don't
      worry if you cannot run these scripts on your machine, your
      patch will be automatically checked in the Actions of the pull
      request.
* [x] I have added an item that describes the changes I have
      introduced in this PR in the section **Changes for next
      release** of the section **Change Control**/**Document history**
      of the document. Create **Changes for next release** if it does
      not exist. Notice that changes that are not modifying the
      content and rendering of the specifications (both HTML and PDF)
      do not need to be listed.
* [X] When modifying content and/or its rendering, I have checked the
      correctness of the result in the PDF output (please refer to the
      instructions on [how to build the PDFs
      locally](https://github.com/ARM-software/acle/blob/main/CONTRIBUTING.md#continuous-integration)).
* [X] The variable `draftversion` is set to `true` in the YAML header
      of the sources of the specifications I have modified.
* [ ] Please *DO NOT* add my GitHub profile to the list of contributors
      in the [README](https://github.com/ARM-software/acle/blob/main/README.md#contributors-) page of the project.

amilendra committed:
Add intrinsics for the FEAT_SVE_AES2 feature introduced by the 2024 dpISA

FEAT_SVE_AES2 adds

1) SVE multi-vector Advanced Encryption Standard (AES) instructions
Instructions added: AESE, AESD, AESEMC and AESDIMC
For each instruction there are two variants
  a) Two registers variant
  b) Four registers variant

2) SVE multi-vector 128-bit polynomial multiply long instructions
Instructions added: PMULL and PMLAL

FEAT_SSVE_AES implements the same instructions but when in streaming mode.

georges-arm commented on main/acle.md:
```diff
@@ -2147,6 +2162,15 @@ support for the SVE2 AES (FEAT_SVE_AES) instructions and if the associated
+ACLE intrinsics are available. This implies that `__ARM_FEATURE_AES`
+and `__ARM_FEATURE_SVE2` are both nonzero.
+
+`__ARM_FEATURE_SSVE_AES2` is defined to 1 if there is hardware support for
```
`__ARM_FEATURE_SSVE_AES2` doesn't appear in the table below, should it?

georges-arm commented on main/acle.md:
```diff
@@ -2569,6 +2623,7 @@ be found in [[BA]](#BA).
 | [`__ARM_BF16_FORMAT_ALTERNATIVE`](#brain-16-bit-floating-point-support)                                                                                 | 16-bit brain floating-point, alternative format                                                    | 1           |
 | [`__ARM_BIG_ENDIAN`](#endianness)                                                                                                                       | Memory is big-endian                                                                               | 1           |
 | [`__ARM_FEATURE_AES`](#aes-extension)                                                                                                                   | AES Crypto extension (Arm v8-A)                                                                    | 1           |
+| [`__ARM_FEATURE_AES2`](#aes-extension)                                                                                                                  | SVE2 Multi-vector AES Crypto extension (Arm v9.6-A)                                                | 1           |
```
`__ARM_FEATURE_AES2` isn't referenced in the "AES Extension" section above?

georges-arm commented on main/acle.md:
```diff
@@ -9405,6 +9466,30 @@ to work with `svboolx2_t` and `svboolx4_t`.  For example:
+  // Variants are also available for:
+  // _s64x2, _f64x2
+  svuint64x2_t svpmull[_u64x2](svuint64_t zn, svuint64_t zm);
+  svuint64x2_t svpmlal[_u64x2](svuint64_t zn, svuint64_t zm);
```
PMLAL is an accumulating instruction, is missing a `svuint64x2_t` input?

georges-arm replied:
My comment was on PMLAL, not PMULL. PMLAL is the eor-accumulating variant of PMULL: https://developer.arm.com/documentation/ddi0602/2025-06/SVE-Instructions/PMLAL--Multi-vector-polynomial-multiply-long-and-accumulate-

georges-arm commented on main/acle.md:
```diff
@@ -9454,6 +9466,30 @@ to work with `svboolx2_t` and `svboolx4_t`.  For example:
+svuint8x2_t    svaesemc[_u8_x2]   (svuint8x2_t op1, svuint64_t op2, uint64_t index);
+svuint8x4_t    svaesemc[_u8_x4]   (svuint8x4_t op1, svuint64_t op2, uint64_t index);
+svuint8x2_t    svaesdimc[_u8_x2]  (svuint8x2_t op1, svuint64_t op2, uint64_t index);
+svuint8x4_t    svaesdimc[_u8_x4]  (svuint8x4_t op1, svuint64_t op2, uint64_t index);
```
All of these instructions are destructive encodings (the output register is also an input register), so we should have an additional parameter for all of these?

georges-arm replied:
Nevermind, I had misread that the Zdn register parameters were duplicated!

AlfieRichardsArm reviewed (commented):
I think this should use `SSVE_AES` instead of `SSVE_AES2`

I believe the two `FEAT`s here are `FEAT_SSVE_AES` and `FEAT_SVE_AES2`. `FEAT_SSVE_AES2` doesn't exist from what I can see.

amilendra committed:
Fix the AES2 review comments.

1.__ARM_FEATURE_SSVE_AES2 -> __ARM_FEATURE_SSVE_AES
2. Move the intrinsics to their own section under SVE2

amilendra committed:
Add SVE2 AES2 support status (Alpha) to the content

amilendra committed:
[PMLAL] Add svuint64x2_t zda which acts as both input/output

amilendra committed:
Address further AES2 review comments

- __ARM_FEATURE_SVE2_AES2 -> __ARM_FEATURE_SVE_AES2
- Add _lane prefix

amilendra committed:
Fix Typos: SVE2 -> SVE

amilendra commented on main/acle.md:
```diff
@@ -9712,6 +9728,42 @@ Lookup table read with 4-bit indices.
+  svuint8x2_t    svaesemc_lane[_u8_x2]   (svuint8x2_t op1, svuint64_t op2, uint64_t index);
+  svuint8x4_t    svaesemc_lane[_u8_x4]   (svuint8x4_t op1, svuint64_t op2, uint64_t index);
+  svuint8x2_t    svaesdimc_lane[_u8_x2]  (svuint8x2_t op1, svuint64_t op2, uint64_t index);
+  svuint8x4_t    svaesdimc_lane[_u8_x4]  (svuint8x4_t op1, svuint64_t op2, uint64_t index);
```
I think that would make it easier to map the parameters to the respective register names in the ISA doc. However existing AES intrinsics follow the current convention.
https://developer.arm.com/architectures/instruction-sets/intrinsics/#q=aes
@CarolineConcatto , @Lukacma what do you think?

CarolineConcatto commented on main/acle.md:
```diff
@@ -9712,6 +9728,42 @@ Lookup table read with 4-bit indices.
+
+  // Variants are also available for:
+  // _s64x2, _f64x2
+  svuint64x2_t svpmull[_u64x2](svuint64_t zn, svuint64_t zm);
```
I believe that we also should have the _n version, like I see in the llvm implementation.
I am proposing something like this:

svuint64x2_t svpmull_u64x2(svuint64_t zn, svuint64_t zm);
svuint64x2_t svpmull[_n]_u64x2(svuint64_t zn, uint64_t zm);

svuint64x2_t svpmlal[_u64x2](svuint64x2_t zda, svuint64_t zn, svuint64_t zm);
svuint64x2_t svpmlal[_n_u64x2](svuint64x2_t zda, svuint64_t zn, uint64_t zm)

amilendra committed:
Add [_n] optional suffix for PMULL, PMLAL intrinsics. Make _u64x2 non-optional for PMULL

amilendra committed:
Match parameter names with instruction operand names

Lukacma committed:
Fix pmull/pmlal intrinsics

Lukacma committed:
remove wrong pmull pmlal variants

CarolineConcatto reviewed (approved):
The prototypes look sensible and are also implemented as is in LLVM.
I would just change the definition of  __ARM_FEATURE_SSVE_AES
to be : Streaming  SVE support for the multi-vector

Lukacma committed:
Adjust feat definition

Lukacma committed:
added full stops

Lukacma committed:
Fix aes zm types

Lukacma committed:
Merge remote-tracking branch 'acle/main' into 2025-acle-sve-aes2

sivan-shani reviewed (approved):
LGTM

vhscampos committed:
Merge branch 'main' into 2025-acle-sve-aes2

sallyarmneale reviewed (commented):
no changes required.
